llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp

```diff
@@ -3018,6 +3018,39 @@ void LibCallSimplifier::classifyArgUse(
   }
 }
 
+/// Constant folds remquo
+Value *LibCallSimplifier::optimizeRemquo(CallInst *CI, IRBuilderBase &B) {
+  const APFloat *X, *Y;
+  if (!match(CI->getArgOperand(0), m_APFloat(X)) ||
+      !match(CI->getArgOperand(1), m_APFloat(Y)))
+    return nullptr;
+
+  if (X->isNaN() || Y->isNaN() || X->isInfinity() || Y->isZero())
+    return nullptr;
```

Comment:
Should be able to handle these easily for not-strictfp

Reply:
See https://en.cppreference.com/w/cpp/numeric/math/remquo:
> If the implementation supports IEEE floating-point arithmetic (IEC 60559),
The current [rounding mode](https://en.cppreference.com/w/cpp/numeric/fenv/FE_round) has no effect.
[FE_INEXACT](https://en.cppreference.com/w/cpp/numeric/fenv/FE_exceptions) is never raised.
If x is ±∞ and y is not NaN, NaN is returned and [FE_INVALID](https://en.cppreference.com/w/cpp/numeric/fenv/FE_exceptions) is raised.
If y is ±0 and x is not NaN, NaN is returned and [FE_INVALID](https://en.cppreference.com/w/cpp/numeric/fenv/FE_exceptions) is raised.
If either x or y is NaN, NaN is returned.

IIRC llvm shouldn't convert these invalid calls into fp exceptions.

Reply:
Well, it could/should if we had a generic raise FP except mechanism but we don't. We do not care about fp exceptions, except in strictfp, and only for calls that aren't fpexcept.ignore. So you should still fold this unless the call is strictfp

Reply:
It is guarded in
https://github.com/llvm/llvm-project/blob/e64e745e8fb802ffb06259b1a5ba3db713a17087/llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp#L3819-L3826